(wip)raft: introduce term cache

hakuuww opened the pull request:
Don't review. Prototype for the termCache. 

More info here https://cockroachlabs.atlassian.net/wiki/x/MgB7AAE

References: cockroachdb#136296
Epic: None
Release note: None

hakuuww committed:
logstore: log latency if raftEntry cache miss when calling loadTerm

This is a draft commit that adds a metric to record the latency for loading term if there is a cache miss.

Also adds separate metrics for raftRntry cache accesses from loadTerm.

Running roachtests against this build will help us with understanding the access patterns of raftEntry Cache, and understand how much performance improvement we may get when we introduce a term cache.

This commit is only for reference purpose, shouldn't be merged.

References: cockroachdb#136296
Epic: None
Release note: None

hakuuww committed:
(wip)raft: introduce term cache

Prototype for the termCache.
And append logic.

References: cockroachdb#136296
Epic: None
Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/141524)
<!-- Reviewable:end -->

hakuuww commented:
slice memory access issues

hakuuww committed:
(wip)raft: introduce term cache

Passed all unit tests in kv/kvserver!

References: cockroachdb#136296
Epic: None
Release note: None

hakuuww committed:
(wip)raft: introduce term cache

Integrate with metrics from another branch #141159

References: cockroachdb#136296
Epic: None
Release note: None

hakuuww commented:
passed all tests in kv server

pav-kv commented on pkg/raft/termCache.go:
```diff
@@ -0,0 +1,309 @@
+
+// Term returns the entry term based on the given entry index
+// Returns error if not in the termCache
+func (tc *TermCache) Term(index uint64) (term uint64, err error) {
```
Probably sufficient to return a bool instead of `error`. Don't need much of a granularity here - just found or not found.

hakuuww replied:
Depends on whether we want the termCache to be a best effort cache.

pav-kv commented on pkg/raft/termCache.go:
```diff
@@ -0,0 +1,309 @@
+	}
+
+	// in last term of termCache
+	if index >= tc.getLastFlipEntry().index && index <= tc.lastIndex {
```
Don't need to check `index <= lastIndex` - already checked this above.

pav-kv commented on pkg/raft/termCache.go:
```diff
@@ -0,0 +1,309 @@
+		return tc.getLastFlipEntry().term, nil
+	}
+
+	for i := 0; i < len(tc.cache)-1; i++ {
+		if index >= tc.cache[i].index && index < tc.cache[i+1].index {
+			return tc.cache[i].term, nil
+		}
+	}
```
Can scan from the back, and halve the slice accesses:

```golang
for i := len(tc.cache)-2; i >= 0; i-- {
	if index >= tc.cache[i].index {
		return tc.cache[i].term
	}
}
```

pav-kv reviewed (commented):
We need to also clear/initialize the cache somewhere in `applySnapshot` stack - the log is cleared there.

pav-kv commented on pkg/raft/termCache.go:
```diff
@@ -0,0 +1,309 @@
+
+// Match returns whether the entryID is in the TermCache.
+// If it is in the termCache, then it is in the raftLog.
+func (tc *TermCache) Match(argEntryId entryID) (bool, error) {
```
Usually, if a function(ality) is not used immediately in the PR, or there is no clear vision on how it will be used in follow-up PRs, it should not be added.

For this `Match` method, do you expect any use? Seems like it's already covered by `raftLog.matchTerm` function, and having the `Term()` method is enough to support it.

pav-kv commented on pkg/raft/termCache.go:
```diff
@@ -0,0 +1,295 @@
+
+// ClearTo clears entries from the TermCache with index strictly less than hi.
+// Mirrors the clearTo function in raftentry/cache.go
+// TODO(hakuuww): actually, we are clearing more than hi here,
+//
+//	which causes TermCache misses in steady state.
```
I think we need to fix this. Truncations are happening pretty much all the time.

2 options:
- if a truncation index is in the middle of `tc.cache[i]`, bump the `tc.cache[i]` to match the truncation index.
- leave the `tc.cache[i]` intact. There is no harm in remembering terms of entries below `firstIndex`.

hakuuww replied:
(from pavel):
---
say you have a log [firstIndex=10] t1 t1 t1 t2 t2 t2
which is then truncated to start at index 12, so the new log is [firstIndex=12] t1 t2 t2 t2
the term cache before the change was: {index=10,term=1}, {index=13,term=2}
I’m saying that, after the truncation, we shouldn’t shrink the cache to be {index=13,term=2}. Because it’s unable to tell us the term of entry 12 that remains in the log.
Options:
make it {index=12,term=1}, {index=13,term=2}
leave it as {index=10,term=1}, {index=13,term=2}
Both options are able to answer the term for entry 12. Option 2 is slightly better because it doesn’t require rewriting the cache. So, looking forward, it will not require updating itself in storage.
[4:55](https://cockroachlabs.slack.com/archives/D08841WBSPJ/p1740088538571509)
it’s ok if the term cache remembers “more” history than the log itself. We wouldn’t be reading terms of entries below firstIndex-1

Inside termCache, we can pretty much assume that the caller has already made the necessary bound checks. The caller won’t be trying to access terms outside the log’s first-1/last index.
---

hakuuww replied:
btw just to be clear **truncateFrom**  is not the problem. its **clearTo** that we are worrying about. (the wording we use is a bit confusing).

truncateFrom is removing suffix of the raftlog. 
clearTo is removing prefix of the raftLog.

hakuuww replied:
Question: why don't we just no do clearTo at all, if it is ok to leave truncated entries in the termCache?

hakuuww commented:
The prototype works as intended and greatly reduces entry cache accesses according to metrics, but there are additional details that we need to figure out. 

1. For the initial prototype, we are truncating and clearing more indices than we necessarily need. 

Originally, I thought **all entries in the `TermCache` must represent term flip points** is an invariant. 

To always preserve this above invariant: 
when we clear some entries up to(not including) a high `hi` index,  we delete the underlying term flip entry with that symbolizes "flipped to the term at `hi`".

(when we truncate entries starting from and including 'lo', all entries below 'lo' are still preserved) 
(when we clear entries up to 'hi', entries at or above 'hi' are preserved, entries below 'hi' are removed) 

This means with an subsequent term look up at the termCache, it no longer thinks that the termCache has entries that are in the same term as in 'hi'. 

Which leads to look ups in the lower level cache(raft entry cache) or storage. 

clearTo operations happen all the time, because the storage engine is always persisting/applying entries. And telling the raftLog to stop storing them. 
This is why we observe termCache misses in metrics.

We can eliminate those termCache misses if we modify the properties of our termCache:
"The `cache[0]` is not necessarily a term flip. All `cache[i]` with i >= 1 are term flips though."

Another reason why we want this has to do with:
1. follower snapshot handling
when the follower receives a snapshot, it would wipe the whole termCache.
the 'lastIndex' of the snapshot is likely not a termFlip, but we still want to add that lastIndex to the termCache at 'termCache[0]'. indicating the log is in term 'termCache[0].term' starting from termCache[0].index up to 'min(cache[i+1].index-1 /* if not nil*/, lastIndex)'

3. always have something in the TermCache even in code restart

hakuuww commented:
TODO: read more go slice low level implementation. We are doing some copying that are not necessary. (ideally a circular array is best, but too overkill for a slice this short?)

hakuuww commented:
Question: why don't we just no do clearTo at all, if it is ok to leave truncated entries in the termCache?

hakuuww committed:
(wip)raft: update clearTo, update entry to last log entry upon processing snapshot and cold restart

References: cockroachdb#136296
Epic: None
Release note: None

hakuuww committed:
(wip)raft: remove unnecessary truncation, add locking for termCache, add
recover from snapshot

References: cockroachdb#136296
Epic: None
Release note: None
